core/src/main/java/com/graphhopper/routing/AStarBidirection.java

```diff
@@ -61,272 +53,95 @@
-     */
-    public AStarBidirection setApproximation(WeightApproximator approx) {
-        weightApprox = new ConsistentWeightApproximator(approx);
-        return this;
-    }
-
-    public WeightApproximator getApproximation() {
-        return weightApprox.getApproximation();
-    }
-
-    @Override
-    protected SPTEntry createSPTEntry(int node, double weight) {
-        throw new IllegalStateException("use AStarEdge constructor directly");
-    }
-
     @Override
-    public void initFrom(int from, double weight) {
-        currFrom = new AStarEntry(EdgeIterator.NO_EDGE, from, weight, weight);
+    void init(int from, double fromWeight, int to, double toWeight) {
         weightApprox.setFrom(from);
-        pqOpenSetFrom.add(currFrom);
-
-        if (currTo != null) {
-            currFrom.weight += weightApprox.approximate(currFrom.adjNode, false);
-            currTo.weight += weightApprox.approximate(currTo.adjNode, true);
-        }
-
-        if (!traversalMode.isEdgeBased()) {
-            bestWeightMapFrom.put(from, currFrom);
-            if (currTo != null) {
-                bestWeightMapOther = bestWeightMapTo;
-                updateBestPath(GHUtility.getEdge(graph, from, currTo.adjNode), currTo, from);
-            }
-        } else if (currTo != null && currTo.adjNode == from) {
-            // special case of identical start and end
-            bestPath.sptEntry = currFrom;
-            bestPath.edgeTo = currTo;
-            finishedFrom = true;
-            finishedTo = true;
-        }
-    }
-
-    @Override
-    public void initTo(int to, double weight) {
-        currTo = new AStarEntry(EdgeIterator.NO_EDGE, to, weight, weight);
         weightApprox.setTo(to);
-        pqOpenSetTo.add(currTo);
-
-        if (currFrom != null) {
-            currFrom.weight += weightApprox.approximate(currFrom.adjNode, false);
-            currTo.weight += weightApprox.approximate(currTo.adjNode, true);
-        }
-
-        if (!traversalMode.isEdgeBased()) {
-            bestWeightMapTo.put(to, currTo);
-            if (currFrom != null) {
-                bestWeightMapOther = bestWeightMapFrom;
-                updateBestPath(GHUtility.getEdge(graph, currFrom.adjNode, to), currFrom, to);
-            }
-        } else if (currFrom != null && currFrom.adjNode == to) {
-            // special case of identical start and end
-            bestPath.sptEntry = currFrom;
-            bestPath.edgeTo = currTo;
-            finishedFrom = true;
-            finishedTo = true;
-        }
+        super.init(from, fromWeight, to, toWeight);
     }
 
     @Override
-    protected Path createAndInitPath() {
-        bestPath = new PathBidirRef(graph, weighting);
-        return bestPath;
+    protected boolean finished() {
+        // using 'weight' is important and correct here e.g. approximation can get negative and smaller than 'weightOfVisitedPath'
+        return super.finished();
```

Comment:
Ok, so actually I think you wrote this comment and the sole purpose of overriding this method here was to have a place where I can put the comment because I did not want to delete it :) 
I think what you wanted to say with this comment is this: For `AStar` we store two weights in the shortest path entries: The first one `weight` is the one that determines the order in which the nodes are visited (the order in the priority queue) and includes the approximate remaining distance to the target. The second one `weightOfVisitedPath` stores the real weight a shortest path entry has (without approximation). The latter is also used for the weight in `bestPath`. When we check if the search has finished we do this

    currFrom.weight + currTo.weight >= bestPath.getWeight()

So we actually compare the weight (with approximation) to the weight without approximation in bestPath and I think the comment is trying to make the reader aware that we use `currFrom/To.weight` not `currFrom/To.weightOfVisitedPath`. However, I do not think I fully understand what is meant with 'approximation can get negative and smaller than weight'.


    `currFrom.weight + currTo.weight >= bestPath.getWeight()`

we use 'weight'

Reply:
> you wrote this comment 

Haha, ok ;)

Very strange comment. I added it in this commit about a performance improvement: https://github.com/graphhopper/graphhopper/commit/37df1b7e381d3e138b424a840b549b51aeae096a

Also this comment in the same commit is strange: https://github.com/graphhopper/graphhopper/commit/37df1b7e381d3e138b424a840b549b51aeae096a#diff-4bdaac1f080eb5164041e7db4280d193L284

Probably this comment should be removed to avoid further confusion.